Add support for Consul Intentions

twunderlich-grapl opened the pull request:
<!-- Thank you for your contribution to Grapl! Please take some time
to fill out this short template to help us review and understand your
PR. -->

### Which issue does this PR correspond to?
https://github.com/grapl-security/issue-tracker/issues/774

<!-- Please provide a link to the GitHub Issue this pull request is
meant to address. -->

### What changes does this PR make to Grapl? Why?
This adds basic consul intentions and the ability to deploy these intentions in a managed manner.

Consul Intentions should be thought of as a service to service firewall. It supports L4 rules (allow or deny traffic from one service to another) and  L7 rules (Allow or deny traffic based on any combination of path, header or method). This will likely become more important as we implement plugins or expand the number of services supported.

Design:
Intentions are currently manually created within the nomad/intentions folder as json files.
Pulumi then takes the directory of these config files reads them in one by one and creates corresponding Consul Intentions dynamically.

Potential future work: Autogenerating stubbed files based on Nomad definitions

<!-- Please describe the functional impact of these changes and the
rationale behind them. This will help us understand your approach to
solving the problem. -->

### How were these changes tested?
Deployed these changes to AWS sandbox and checked that intentions were created in consul

<!-- Please describe how these changes were tested. There should be
sufficient detail that reviewers can replicate your testing
procedure. If the procedure is a manual one that's OK, your
description will help us work with you to get it automated. -->

Commit:
Install new requirements.

Commit:
Add consul intention setup

Commit:
Switch to json format for intention files

Commit:
Remove hcl parsing library since we're now using json files

Commit:
Remove hcl intention files

Commit:
Update README

Commit:
Add support for connecting to pulumi via consul stack and related cleanup

Commit:
Add back the require address

twunderlich-grapl commented on nomad/intentions/dgraph-alpha-0-grpc-private.json:
```diff
@@ -0,0 +1,12 @@
+    {
+      "Name": "dgraph-zero-0-grpc-private",
+      "Action": "allow",
+      "Precedence": 9,
```
We add the precedence and  Type here so that pulumi doesn't always diff. There are the defaults for consul intentions

Commit:
Formatting of course

Commit:
It turns out that for the tests we apparently do need to set consul:address to the docker host network.

christophermaier commented on pulumi/infra/consul_intention.py:
```diff
@@ -0,0 +1,32 @@
+                    resource_name=f"{name}-{intention['Name']}",
+                    kind=intention["Kind"],
+                    name=intention["Name"],
+                    config_json=json.dumps({"Sources": intention["Sources"]}),
```
It's super weird that the provider doesn't just accept HCL directly 🤔

twunderlich-grapl replied:
Agreed. I originally wrote the conf files in hcl for better readability before realizing that the consul provider can't take hcl and that the config has to be in json format instead of a dict. Very weird api design in general by the terraform module.

christophermaier commented on pulumi/grapl/__main__.py:
```diff
@@ -159,6 +160,11 @@ def main() -> None:
+        ConsulIntention(
+            "grapl-core",
+            intention_directory=Path("../../nomad/intentions").resolve(),
+        )
```
As I'm reading this code, the fact that the class is called `ConsulIntention`, but the intentions themselves are stored in a `nomad` directory makes me think that this is somehow a copy/paste error. If I were looking for where our Consul intentions are defined, I would instinctively look for a `consul` directory; it would take a while before I thought to look for them in a `nomad` directory, even though they are intentions that govern the behavior of our Nomad-managed services.

One solution would be to rename the directory `consul`. Another could be to rename `intentions` to `consul-intentions`.

On top of either of those, it may also be worth renaming `ConsulIntention` to something like `GraplServiceIntentions`.

twunderlich-grapl replied:
Added a comment regarding not needing an explicit provider. Basically its the same reason that the NomadJobs in local don't need one. Pulumi will use the consul:address if its set in the config or use the default localhost:8500 address. We need an explicit provider for the AWS section because we're setting the address from the StackReference in some cases.

twunderlich-grapl replied:
> As I'm reading this code, the fact that the class is called `ConsulIntention`, but the intentions themselves are stored in a `nomad` directory makes me think that this is somehow a copy/paste error. If I were looking for where our Consul intentions are defined, I would instinctively look for a `consul` directory; it would take a while before I thought to look for them in a `nomad` directory, even though they are intentions that govern the behavior of our Nomad-managed services.
> 
> One solution would be to rename the directory `consul`. Another could be to rename `intentions` to `consul-intentions`.
> 
> On top of either of those, it may also be worth renaming `ConsulIntention` to something like `GraplServiceIntentions`.

I've gone down the route of renaming the existing folder to consul-intentions. I believe that if we want people to remember to update the intention rules, it'll be easier if the config files are colocated. Medium-term I'd want to add enforcement and autogeneration to make this easier.

I also think that its worth keeping the explicit `ConsulIntentions` name so that new hires know that `intentions` are a consul construct and not a grapl-specific one.

christophermaier commented on pulumi/infra/consul_intention.py:
```diff
@@ -0,0 +1,32 @@
+            with open(file, "r") as f:
+                intention = json.load(f)
+                consul.ConfigEntry(
+                    resource_name=f"{name}-{intention['Name']}",
```
It might be a good idea to add "-intention" to the end of this name, given that a `ConfigEntry` can be more than just an intention.

christophermaier commented on pulumi/infra/consul_intention.py:
```diff
@@ -0,0 +1,32 @@
+                intention = json.load(f)
+                consul.ConfigEntry(
+                    resource_name=f"{name}-{intention['Name']}",
+                    kind=intention["Kind"],
```
It might also be useful to enforce that the incoming `Kind` is, indeed, "service-intentions". You are implicitly assuming this with `intention["Sources"]` below (unless some other Consul configuration entry also has a `Sources` field); beyond the assumption of a "Sources" field, nothing in this class is actually intention-specific.

(You could check for both of these conditions immediately after the `json.load` call above, and provide an error message if the conditions aren't met.)

twunderlich-grapl commented on src/python/mypy.ini:
```diff
@@ -30,4 +30,7 @@ ignore_missing_imports = True
\ No newline at end of file
+ignore_missing_imports = True
+
+[mypy-hcl2]
```
This is a holdover from a previous iteration. Removing now.

Commit:
FIrst round of updates based on feedback
- Add error checking for consul intentions with nice error messages
- Add parent-child link to intentions
- Set name to be explicitly an intention
- Remove last hcl2 library reference
- Readme grammar fix

Commit:
Fix typing

Commit:
Dry out getting a custom provider with address set

Commit:
Rename to make the plural more explicit

Commit:
TIL that mypy doesn't support module subclassing Protocols

Commit:
Allow pulumi to reach consul

Commit:
Fix lint checks

twunderlich-grapl commented:
@christophermaier this is ready for re-review

twunderlich-grapl commented on pulumi/infra/get_hashicorp_provider_address.py:
```diff
@@ -0,0 +1,24 @@
+    address = override_address or stack.require_output(
+        "address"
+    )
+    return pulumi_class.Provider(f"{provider_type}-aws", address=address)
```
I was using it to specify that this was an address in aws, but its unnecessary. I'll remove it

Commit:
Remove aws suffix from provider pulumi name
